src/lib.rs

```diff
@@ -0,0 +1,160 @@
+use lazy_static::lazy_static;
+
+use guest::prelude::*;
+use kubewarden_policy_sdk::wapc_guest as guest;
+
+use k8s_openapi::api::core::v1 as apicore;
+
+extern crate kubewarden_policy_sdk as kubewarden;
+use kubewarden::{logging, protocol_version_guest, request::ValidationRequest, validate_settings};
+
+mod settings;
+use settings::Settings;
+
+use slog::{o, warn, Logger};
+
+lazy_static! {
+    static ref LOG_DRAIN: Logger = Logger::root(
+        logging::KubewardenDrain::new(),
+        o!("policy" => "sample-policy")
+    );
+}
+
+#[no_mangle]
+pub extern "C" fn wapc_init() {
+    register_function("validate", validate);
+    register_function("validate_settings", validate_settings::<Settings>);
+    register_function("protocol_version", protocol_version_guest);
+}
+
+fn mutate_request(pvc: apicore::PersistentVolumeClaim, settings: &Settings) -> CallResult {
+    warn!(
+        LOG_DRAIN,
+        "storage class \"{}\" is not allowed, using fallback storage class",
+        settings.fallback_storage_class.clone().unwrap_or_default()
+    );
+    let mutated_pvc = serde_json::to_value(apicore::PersistentVolumeClaim {
+        spec: Some(apicore::PersistentVolumeClaimSpec {
+            storage_class_name: settings.fallback_storage_class.clone(),
+            ..pvc.spec.unwrap_or_default()
+        }),
+        ..pvc
+    })?;
+    kubewarden::mutate_request(mutated_pvc)
+}
+
+fn validate(payload: &[u8]) -> CallResult {
+    let validation_request: ValidationRequest<Settings> = ValidationRequest::new(payload)?;
+
+    // TODO: you can unmarshal any Kubernetes API type you are interested in
+    match serde_json::from_value::<apicore::PersistentVolumeClaim>(
+        validation_request.request.object,
+    ) {
+        Ok(pvc) => {
+            // TODO: your logic goes here
+            if let Err(e) = has_valid_storage_class(&pvc, &validation_request.settings) {
+                if validation_request.settings.fallback_storage_class.is_some() {
+                    return mutate_request(pvc, &validation_request.settings);
+                }
+                return kubewarden::reject_request(Some(e), None, None, None);
+            }
+            kubewarden::accept_request()
+        }
+        Err(_) => {
+            // TODO: handle as you wish
+            // We were forwarded a request we cannot unmarshal or
+            // understand, just accept it
+            warn!(LOG_DRAIN, "cannot unmarshal resource: this policy does not know how to evaluate this resource; accept it");
+            kubewarden::accept_request()
+        }
+    }
+}
+
+fn has_valid_storage_class(
+    pvc: &apicore::PersistentVolumeClaim,
+    settings: &Settings,
+) -> Result<(), String> {
```

Comment:
I don't like the usage of an error, what about doing:

```rust
fn uses_denied_storage_class(pvc: PVC, settings: &Settings) -> (bool, Option<String>)
```

bool is about using a storage class that is on the deny list.

* A good storageclass is used -> `(false, None)`
* A bad storageclass is used -> `(true, Some("name"))`